crates/matrix-sdk-common/src/linked_chunk/mod.rs

```diff
@@ -528,6 +528,46 @@ impl<const CAP: usize, Item, Gap> LinkedChunk<CAP, Item, Gap> {
         Ok(removed_item)
     }
 
+    /// Replace item at a specified position in the [`LinkedChunk`].
+    ///
+    /// `position` must point to a valid item, otherwise the method returns
+    /// `Err`.
+    pub fn replace_item_at(&mut self, position: Position, item: Item) -> Result<(), Error>
+    where
+        Item: Clone,
+    {
+        let chunk_identifier = position.chunk_identifier();
+        let item_index = position.index();
+
+        let chunk = self
+            .links
+            .chunk_mut(chunk_identifier)
+            .ok_or(Error::InvalidChunkIdentifier { identifier: chunk_identifier })?;
+
+        match &mut chunk.content {
+            ChunkContent::Gap(..) => {
+                return Err(Error::ChunkIsAGap { identifier: chunk_identifier })
+            }
+
+            ChunkContent::Items(current_items) => {
+                if item_index > current_items.len() {
+                    return Err(Error::InvalidItemIndex { index: item_index });
+                }
+
+                if let Some(updates) = self.updates.as_mut() {
+                    updates.push(Update::ReplaceItem {
+                        at: Position(chunk_identifier, item_index),
+                        item: item.clone(),
+                    });
+                }
```

Comment:
Push the update after `current_items[item_index] = item;`: do the action before triggering the update.

Reply:
This means we have to clone *before*, and the original thing is unused if `self.updates.as_mut().is_none()`. Plus, the updates aren't reflected in real-time, they're buffered until they're consumed, as far as I can tell? So this should be fine to keep as is. I can add a comment explaining that this avoids a clone if you'd like.